### ISSUE TITLE
Overriding getStateForAction in inner Navigator throws error.

### ISSUE THREAD

**Issue by paulpopiel** [react-navigation/react-navigation#4491]
### Current Behavior

When trying to customize navigation behaviour through overriding the getStateForAction method on an inner Navigator - throws an error.

Error:
TypeError: this.getInitialState is not a function

The offending line is in SwitchRouter.js
```
    getStateForAction(action, inputState) {
      let prevState = inputState ? { ...inputState } : inputState;
      let state = inputState || this.getInitialState();
      let activeChildIndex = state.index;
```

### Expected Behavior

This error was not present prior to v2.

### How to reproduce

https://snack.expo.io/SyyD5yClm

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 2.3.0
| react-native     | 0.55.4
| node             |
| npm or yarn      |

**Comment by brentvatne** [comment]
@ericvicenti - any idea?

**Comment by brentvatne** [comment]
`const defaultGetStateForAction = InnerNavigator.router.getStateForAction.bind(InnerNavigator.router);` will fix it. we should automatically bind this

**Comment by paulpopiel** [comment]
Thanks!

**Comment by brentvatne** [comment]
we don't support 1.x anymore, you will need to backport the fix manually in a fork, sorry!

**Comment by brentvatne** [comment]
but also in the original comment:

> This error was not present prior to v2.

so it should not be a problem?

**Comment by frankiesardo** [comment]
It is, I'm using 1.6 as we speak. I've backported it as this:

https://gist.github.com/frankiesardo/d34a16a40b3c74a465a4aebcc87b5a6a

But I'm not confident this is the correct solution. Shame 1.x is not supported anymore. 2.X and 3.X have introduced too many regression/quirks for my app to update. I was using 2.2.5 because of this problem https://github.com/react-navigation/react-navigation/issues/4481
But then found the fix for new iPhones was only released for 2.15 https://github.com/react-navigation/react-navigation/issues/4985
So I've returned to 1.6 as it is the most stable but if new fixes are not going to be ported back we're in a pickle :(

I understand there is a huge work involved in maintaining all these versions and I am nonetheless grateful for your work.
